### ISSUE TITLE
December 2022 release + 32-bit Raspbian Buster results in constant errors when creating image files of 2GB or larger in size

### ISSUE THREAD

**Issue by stinkerton18** [PiSCSI/piscsi#1057]
# Info

- Which version of Pi are you using: RPi 4B, 3B, 3A
- Which github revision of software: 22.12.1 on 32-bit Rasbian 10 "Buster", ext4 filesystem
- Which board version: v2.3 full-spec
- Which computer is the PiSCSI connected to: PB G3 Lombard, PM 6500, unconnected

After updating to the December 2022 release on a Buster install, I have run into the following issues:

1. Through the web interface, I cannot rename any image files 2GB or larger in size
2. Through the web interface, I cannot create new images, of any kind, with a size of 2GB or larger

Since I was running an older release (which was prior to rebranding), I tried using easyinstall.sh to update, then stopping the rascsi/rascsi-web services, removing their entries from /etc/systemd, removing the /etc/nginx/sites-available/default file, removing the /home/pi/RASCSI folder and recloning the source repo, then running easyinstall.sh to compile/install again.

Issue 1 Details: 
When you try to create a drive image of 2GB or larger, the web interface returns:
"Can't create image file '/home/pi/images/IMAGE--NAME': filesystem error: cannot resize file: Invalid argument [/home/pi/images/IMAGE--NAME]"
Logging in via local console/SSH and running dd to manually create an image works without error and the image shows up to be mounted.

Issue 2 Details:
When you try to rename an existing image of 2GB or larger, the web interface returns a white screen with only the following text:
Internal Server Error
The PiSCSI Web Interface did not get a valid response from PiSCSI. Please go back and try again. If the issue persists, please report a bug.
Logging in via local console/SSH, the images/files can be "mv"ed without error.

Both errors above also causes any currently mounted images to be force dismounted, revering back to what is in default.json. In talking/working with dmark in the troubleshooting channel in Discord, the suspicion is the web interface is running into this same problem: https://raspberrypi.stackexchange.com/questions/82261/value-too-large-for-defined-data-type-error-on-files-larger-than-2gb-only-in-r

After using a new SD card with 64-bit Raspbian 11 "Bullseye", the issues above are no reproducible.

**Comment by stinkerton18** [comment]
Currently working on making a ready-to-run image file of the Buster/December 22 release where these issues are 100% reproducible.

**Comment by stinkerton18** [comment]
As requested by dmark, currently running easyinstall.sh with the `--with_gcc` flag. Once completed, will repeat the tests of the above.

**Comment by stinkerton18** [comment]
Trying to run easyinstall.sh, with `--with_gcc` is not going anywhere. It hangs on the following:
`g++ -O3 -Wall -Werror -Wextra -DNDEBUG -Wno-psabi -std=c++17 -iquote . -D_FILE_OFFSET_BITS=64 -D_LARGEFILE64_SOURCE -MD -MP  -DCONNECT_TYPE_FULLSPEC -c ./piscsi/piscsi_core.cpp -o obj/fullspec/piscsi_core.o`

CPU usage is minimal but wait states are through the roof:
`top - 21:44:27 up  1:57,  1 user,  load average: 6.92, 5.30, 5.51
Tasks: 127 total,   1 running, 126 sleeping,   0 stopped,   0 zombie
%Cpu(s):  0.1 us,  3.4 sy,  0.0 ni, 17.4 id, 79.1 wa,  0.0 hi,  0.1 si,  0.0 st
MiB Mem :    429.0 total,     23.5 free,    392.6 used,     12.9 buff/cache
MiB Swap:    100.0 total,      0.0 free,    100.0 used.      5.8 avail Mem`

**Comment by rdmark** [comment]
You may be running out of RAM. Try using less cores with the —j option.

**Comment by stinkerton18** [comment]
Rerunning with `-j=1` now

**Comment by stinkerton18** [comment]
Compilation completed and installed, services started up as expected. Unfortunately, the issue is reproducible.

**Comment by uweseimet** [comment]
@stinkerton18 Can you please try whether branch fix_issue_1057 works for you? It looks as if the outdated buster gcc and clang cannot even rename big files.
There were other file size issues with these compilers before. Unfortuntately the only work-around is using non-portable code, i.e. code that is not part of the standard C++ libraries. For the two operations you mentioned standard C++ library code is used since release 22.12, which explaines why you observed issues in this release.

**Comment by stinkerton18** [comment]
Can do! I'll start it here very shortly.

I can also confirm I did not observe this issue with that same Rasbian 10 system before, having used it for quite some time now. I have not tried Raspbian 11 32-bit but assume, since 64-bit did not have an issue and the filesystem wasn't the limiting factor, the issue wouldn't be present in Raspbian 11 32-bit either.

**Comment by uweseimet** [comment]
@stinkerton18 Please double-check. https://github.com/PiSCSI/piscsi/branches shows that fix_issue_1057 is an available branch. Run "git pull" and then "git checkout fix_issue_1057".

**Comment by stinkerton18** [comment]
Ahh my mistake. I thought fix_issue_1057 was a specific commit/PR of the develop branch, not a branch of its own. Corrected, pulled, compiling now.

**Comment by stinkerton18** [comment]
We're making progress. I can create image files of 2GB+ in size now but renaming still results in the Web interface throwing an error and all mounted images becoming dismounted.

**Comment by uweseimet** [comment]
@stinkerton18 Unless I am missing something the code for renaming in the test branch is the same as it was in previous releases.
Can you please check whether renaming with scsictl also fails? The command format is:
```
scsictl -R OLD_NAME:NEW_NAME
```
If it fails please provide the error message displayed by scsictl.

**Comment by stinkerton18** [comment]
It does indeed fail:
`pi@RaSCSI-EXT:~/images $ scsictl -R OS_7.5.hda OS_7.5-68K.hda
Error: Invalid file descriptor 'OS_7.5.hda', format is CURRENT_NAME:NEW_NAME`

I can rename the file using regular OS tools:
`pi@RaSCSI-EXT:~/images $ mv OS_7.5.hda OS_7.5-68K.hda
pi@RaSCSI-EXT:~/images $ ls
After_Dark_Deluxe.iso      MacOS_71.iso  OS_7.1.hda      RaSCSI-BootstrapV3.hda
Apple_Legacy_Recovery.iso  MacOS_75.iso  OS_7.5-68K.hda
pi@RaSCSI-EXT:~/images $`

**Comment by stinkerton18** [comment]
That's what I get for multitasking.... Here's the actual error returned:
`pi@RaSCSI-EXT:~/images $ scsictl -R OS_7.5.hda:OS_7.5-68K.hda
Error: Invalid protobuf message header
pi@RaSCSI-EXT:~/images $ ls
After_Dark_Deluxe.iso      MacOS_71.iso  OS_7.1.hda  RaSCSI-BootstrapV3.hda
Apple_Legacy_Recovery.iso  MacOS_75.iso  OS_7.5.hda`

It is returning a protobuffer error.

**Comment by uweseimet** [comment]
I found an SD card with buster somewhere and can now reproduce the issue. https://www.mjr19.org.uk/sw/inodes64.html has more details. Unfortunately almost any filesystem operation with files bigger than 2 GB is concerned, even those only dealing with the file attributes or filenames.
The actual error is piscsi crashing, resulting in a protobuf protocol error on the scsictl side:
```
>./bin/fullspec/scsictl -C test_ibm.hds:x.hds
terminate called after throwing an instance of 'std::filesystem::__cxx11::filesystem_error'
  what():  filesystem error: symlink_status: Value too large for defined data type [/home/us/images/test_ibm.hds]
Aborted
```
Reverting to the old (v22.10.01) code is not really a viable solution, because this code does not deal with some special image file cases.

IMO we should say: Image files > 2GB require release 22.10.01 or a 64 bit OS. bullseye 32 bit might also work, provided that the compiler toolchain supports 64 bit inodes (see the link above), but I cannot test this. @stinkerton18 Maybe you can test 32 bit bullseye?

**Comment by stinkerton18** [comment]
I certainly can test on 32-bit Bullseye. If you want a perfect copy of my image, see my first comment above. There's a couple links to a DD/imageUSB file of my 32GB SD card and Buster install.

**Comment by uweseimet** [comment]
@stinkerton18 Would be great if you could test with 32-bit Bullseye. We would learn which OS (buster/bullseye, 32/64 bit) is actually affected.

**Comment by stinkerton18** [comment]
Bullseye installed, piscsi installed, currently copying over the images I had before via USB, should have results in 20 or so minutes.

**Comment by stinkerton18** [comment]
Ok, can confirm this issue is specific to 32-bit Buster. On a fresh 32-bit Bullseye install, fresh December 22 PiSCSI pull (not the fix_issue_1057 branch), neither issue is present/cannot be reproduced.

**Comment by uweseimet** [comment]
In this case my suggested solution is: Image files > 2 GB on a 32-bit platform require Bullseye and are not supported by Buster.

**Comment by stinkerton18** [comment]
Or just make Bullseye the minimum requirement for any PiSCSI version from December 22 on. IIRC, that build is the first one that fully completes the rebranding anyways, making it probably the cleanest "rebase" point. I recognize it's not my call/decision to make but that's my suggestion.

**Comment by rdmark** [comment]
32bit Buster was the default OS for official release images up until the Oct 2021 release. Given the kinds of feedback we regularly get from users (including @stinkerton18 :) ) we still have an active user base on this particular OS configuration.

The issue does not impact 32bit Bullseye. This is the primary OS that I use with PiSCSI.

@uweseimet You mentioned that handling of some special image file cases led to this breakage. Can you please describe on a high level what those were, and which code changes that were involved? If we go down this path I'd like to be able to explain to the community why we drop support for 32bit Buster.

**Comment by uweseimet** [comment]
These special cases are source or destination files that are not regular files but symbolic links, for instance. 22.12.01 can deal with this to a certain degree, definitely better than older releases. And the filesystem code in 22.12.01 is more portable than before.
The current problem is not so much 32-bit Buster support in general, but for files > 2 GB. Support in general may become a problem in a couple of months, when Bookworm (successor of Bullseye) will be released. Who will spend time on testing outdated OS releases? It takes a user less time to upgrade than a developer to maintain and test old releases.

**Comment by rdmark** [comment]
I agree that we should not keep testing and fixing issues for a never ending long tail of legacy OS versions. If you recall, what we have agreed on in this project is to maintain support for two latest stable Raspbian OS releases. Once Bookworm is released later this year we can stop caring about Buster altogether. If you think this is too long a support period for the older OS version we can revisit this decision.

IMHO, I think it is non-trivial and risky to ask less technical users to do an in-place distribution upgrade. I've seen this go terribly wrong sometimes. The more realistic approach is to ask them to start over with a fresh release image, backing up any local configurations and image files to another computer, which is a fairly cumbersome process. Not to discount the significant effort that developers have to put in to maintain compatibility of course. It's a judgment call that we have to make as a team.

For this particular case, I'm going to put up a PR for the Web Interface to revert back to the previous use of `dd` and Python file operations for image file management. We already have all the methods in the FileCmds class for other purposes so it should be a trivial change.

**Comment by uweseimet** [comment]
@rdmark I was not aware of the support policy (two latest stable releases), but this sounds like a good approach.
Regarding file operations if you use Python you should ensure that your Python code deals correctly with cases like the source file being a symlink or a device file. dd might not work in all cases, or not as expected. Just as before.

**Comment by rdmark** [comment]
I see, somehow I was sure we had made this explicit in a ticket discussion somewhere where you were present, but I may be mistaken. Anyhow, I updated the first paragraphs in https://github.com/PiSCSI/piscsi/wiki/Setup-Instructions to make this support timeline clear.

We will only use dd / fallocate for creating new files so I don't think symlink or device files are a consideration, unless I'm mistaken. I will test the corner cases when operating on existing files.
